clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -128,23 +137,99 @@ record_filter_t::parallel_shard_error(void *shard_data)
     return per_shard->error;
 }
 
+bool
+record_filter_t::write_trace_entry(per_shard_t *shard, const trace_entry_t &entry)
+{
+    if (!shard->writer->write((char *)&entry, sizeof(entry))) {
+        shard->error = "Failed to write to output file " + shard->output_path;
+        success_ = false;
+        return false;
+    }
+    ++shard->output_entry_count;
+    return true;
+}
+
 bool
 record_filter_t::parallel_shard_memref(void *shard_data, const trace_entry_t &entry)
 {
     per_shard_t *per_shard = reinterpret_cast<per_shard_t *>(shard_data);
+    ++per_shard->input_entry_count;
     bool output = true;
+    bool all_stop = true;
     for (int i = 0; i < (int)filters_.size(); ++i) {
-        if (!filters_[i]->parallel_shard_filter(entry, filter_shard_data_[i]))
+        bool stop = false;
+        if (!filters_[i]->parallel_shard_filter(entry, per_shard->filter_shard_data[i],
+                                                &stop))
             output = false;
+        if (stop)
+            per_shard->filter_stop[i] = true;
+        all_stop = all_stop && per_shard->filter_stop[i];
+    }
+
+    // Everything before the timestamp marker in the first unit header is part of the
+    // trace shard's header.
+    if (entry.type == TRACE_TYPE_MARKER && entry.size == TRACE_MARKER_TYPE_TIMESTAMP) {
+        per_shard->in_shard_header = false;
+    }
+
+    // We output everything in the trace shard's header.
+    if (per_shard->in_shard_header) {
+        output = true;
+    }
+
+    // Keep track of the last omitted unit header so that we can output it if and when
+    // we output a trace_entry_t from this unit.
+    if (!output && entry.type == TRACE_TYPE_MARKER) {
+        switch (entry.size) {
+        case TRACE_MARKER_TYPE_TIMESTAMP:
+            // No need to remember the previous unit's header anymore. We're in the
+            // next unit now.
+            per_shard->last_filtered_unit_header.clear();
+            ANNOTATE_FALLTHROUGH;
+        case TRACE_MARKER_TYPE_WINDOW_ID:
+        case TRACE_MARKER_TYPE_CPU_ID:
+            per_shard->last_filtered_unit_header.push_back(entry);
+        }
+    }
+
+    // Since we're outputing something from this unit, output its header.
+    if (output) {
+        for (trace_entry_t &unit_header_entry : per_shard->last_filtered_unit_header) {
+            if (!write_trace_entry(per_shard, unit_header_entry))
+                return false;
+        }
+        per_shard->last_filtered_unit_header.clear();
+    }
+
+    if (!output) {
+        if (entry.type == TRACE_TYPE_FOOTER ||
+            (entry.type == TRACE_TYPE_MARKER &&
+             entry.size == TRACE_MARKER_TYPE_CHUNK_FOOTER)) {
+            output = true;
```

Comment:
If this is not allowed it should be an error rather than silently ignoring what the tool wants?  Or is it better to assume the tool writer knows what they're doing and let them create a seemingly-invalid trace?  IMHO either of those is preferable to silently ignoring requests.